### PR TITLE
Improve redirects

### PR DESCRIPTION
**Pull request by djanowski**
* Follow up to 10 redirects.
* Only follow redirects for 301, 302, 303 and 307 status codes.
* Follow redirects only within same domain.

### CODE REVIEW

File: lib/healthchecks.js [hunk 1]
```diff
@@ -134,8 +136,15 @@ function get(url, headers, resolver, timeout) {
         if (response.statusCode < 200 || response.statusCode >= 400)
           resolve({ statusCode: response.statusCode, body: null, elapsed: elapsed });
         else if (response.statusCode > 300 && response.statusCode < 400) {
```
Comment (assaf) [inline lib/healthchecks.js]: Note that only 301/302/307 are redirects that we should follow.

File: lib/healthchecks.js [hunk 2]
```diff
@@ -121,6 +121,8 @@ function get(url, headers, resolver, timeout) {
     })
   };
 
+  state = state || { redirects: 0 };
```
Comment (assaf) [inline lib/healthchecks.js]: What's in state? What's outside of state?

Comment (djanowski) [inline lib/healthchecks.js]: It was a symptom of the ugly solution I implemented at the beginning using recursion. Upgraded to an async loop :sunglasses:

File: lib/healthchecks.js (hunk 1 above)
Comment (djanowski) [inline lib/healthchecks.js]: Done (added 303 See Other).

**Comment by djanowski** [comment]
@assaf What do you think about elapsed time in the case of redirects? Should we report the sum of all timings? It doesn't matter and just report the timing of last request?

**Comment by assaf** [comment]
Elapsed time is for the check, so it's all inclusive.

File: lib/healthchecks.js [hunk 3]
```diff
@@ -107,20 +114,55 @@ Outcome.prototype.toString = function() {
...
+        })
+      };
+
+      get(request, timeout)
+        .then(function(res) {
+          if (~REDIRECT_STATUSES.indexOf(res.statusCode)) {
+            if (redirects < MAX_REDIRECT_COUNT) {
+              const redirectURL = resolver(res.headers.location);
+              if (withinSameDomain(redirectURL, loopbackURL))
+                queue.emit('request', res.headers.location, redirects + 1);
```
Comment (assaf) [inline lib/healthchecks.js]: How is this different from just calling the function directly?

**Pull request merged**